### PR TITLE
[refactor] Refactor `PrettyPrinter` and enable colors in the general logger

### PR DESCRIPTION
**Pull request by vkarak**
- `PrettyPrinter` is now a proxy to the ReFrame's logger and takes particular care in formatting specially the progress output.
- The general framework logger is now also capable of colorizing its output. This allows errors and warnings to be colorized when logged from anywhere in the framework. It also prefixes errors and warnings with the program name.

This is useful for #623.

### CODE REVIEW

File: reframe/utility/__init__.py [hunk 1]
```diff
@@ -310,3 +311,47 @@ def __eq__(self, *args, **kwargs):
...
+
+
+class Colorizer(abc.ABC):
+    """Base class for colorizing strings on the terminal."""
+
+    def __repr__(self):
+        return debug.repr(self)
+
+    @abc.abstractmethod
+    def colorize(string, foreground, background):
```
Comment (teojgo) [inline reframe/utility/__init__.py]: Should this be also `@staticmethod`?

Comment (vkarak) [reply reframe/utility/__init__.py]: You mean `@classmethod`? Yes, you have a point.

Comment (teojgo) [reply reframe/utility/__init__.py]: No I mean the `@staticmethod` since you don't need access to the class itself.

Comment (teojgo) [reply reframe/utility/__init__.py]: Actually yes, the `@classmethod` would be even better to access the corresponding escape sequences, etc.

Comment (vkarak) [reply reframe/utility/__init__.py]: I'll take a look and fix it.

Comment (vkarak) [reply reframe/utility/__init__.py]: I refactored completely the coloring code.

**Comment by victorusu** [comment]
@vkarak and @teojgo, Can I merge?

**Comment by vkarak** [comment]
@victorusu Yeah, sure! I wanted to do that yesterday.

**Pull request merged**